File: Day 15/Day_15.py
import time


class MemGame:
    def __init__(self, numbers: list):
        self.memory = {}
        self.turn_count = 1
        for number in numbers:
            self.memory[number] = self.turn_count
            self.previous_num = number
            self.turn_count += 1
        del self.memory[self.previous_num]  # Don't want to remember the last value quite yet
        print(self.memory)

    def process_turn(self):
        previous_turn = self.turn_count - 1
        if self.previous_num in self.memory:
            _next = previous_turn - self.memory[self.previous_num]
        else:
            _next = 0

        self.memory[self.previous_num] = previous_turn
        self.turn_count += 1
        self.previous_num = _next

    def process_to(self, stop: int):
        """Note: the stop condition works the same way slices do, so to go up to and including 2020, set stop to 2021"""
        while self.turn_count < stop:
            self.process_turn()
        print(f"Turn {self.turn_count - 1}: {self.previous_num}")


def parse_input(filename):
    with open(filename, "r") as file:
        return [int(i) for i in file.read().strip().split(",")]


def main():
    start_time = time.time()
    numbers = parse_input("input.txt")
    game = MemGame(numbers)
    game.process_to(2021)
    game.process_to(30000001)
    print(game.previous_num)
    print("Elapsed time:", time.time() - start_time)


if __name__ == "__main__":
    main()

File: Day 15/Day 15.cpp
#include <vector>
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>
#include <cctype>

#include "stringsplit.h"
#include "timer.h"

class MemGame {
public:
	std::vector<int> memory;  // too big for std::array
	int turn_count;
	int previous_num;
	MemGame(std::vector<int> numbers) 
		: memory(30000000, -1), previous_num(0)
	{
		turn_count = 1;
		for (const auto& number : numbers) {
			memory[number] = turn_count;
			previous_num = number;
			turn_count++;
		}
		memory[previous_num] = -1;
	}

	void process_turn() {
		int next;
		int previous_turn = turn_count - 1;
		if (memory[previous_num] == -1) {  // wasn't found
			next = 0;
		}
		else next = previous_turn - memory[previous_num];

		memory[previous_num] = previous_turn;
		turn_count++;
		previous_num = next;
	}
	
	// Note that we stop at this turn without doing it, so if you want to run through turn 20, you need to specify 21.
	void process_to(int stop) {
		while (turn_count < stop)
			process_turn();
		std::cout << "Turn " << turn_count - 1 << ": " << previous_num << "\n";
	}
};

std::vector<int> parse_input(const char* filename) {
	std::vector<int> output;
	std::ifstream file(filename);
	std::stringstream raw_contents;
	raw_contents << file.rdbuf();
	auto raw_entries = split(strip(raw_contents.str()), ",");
	for (const auto& entry : raw_entries) {
		output.push_back(std::stoi(entry));
	}
	return output;
}

int main() {
	Timer timer;
	auto numbers = parse_input("input.txt");
	MemGame game(numbers);
	game.process_to(2021);
	game.process_to(30000001);
}